### PR TITLE
FPV-44 Integrate Amazon S3 changes into candidate routes

### PR DESCRIPTION
**Pull request by kendrick-807**
Updates & deletes the image stored in s3 as well as the data stored in mongoDB

@phamduylong @jaakkoiot @Niklasnk @kikani-parth

### CODE REVIEW

**Review by nickyous — APPROVED** [review]
Approved

File: routes/candidateRoutes.js [hunk 1]
```diff
@@ -85,33 +85,49 @@ candidateRouter.post('/create', authorizedAdmin, upload.single('file'), async (r
...
             const imageSaved = await s3.send(command);
             if(!imageSaved){
+                logger.error("Unable to upload image to s3");
                 return res.status(400).send({error: 'Unable to upload image'});
             }
             logger.info("Image created successfully! Details:", imageSaved);
             const candidate = new Candidate({name: name, age: age, message: message, img: `https://${BUCKET_NAME}.s3.amazonaws.com/${name}_${currentTimestamp}`});
             const candidateSaved = await candidate.save();
-            if(candidateSaved) {
+            if(candidateSaved && duplicateCandidate.length === 0) {
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: Will simply never hit. Condition guaranteed to be false.

File: routes/candidateRoutes.js [hunk 2]
```diff
@@ -85,33 +85,49 @@ candidateRouter.post('/create', authorizedAdmin, upload.single('file'), async (r
...
                 logger.info("Candidate created successfully! Details:", candidateSaved);
                 return res.status(200).send({ message: "Candidate created successfully!" });
+            }else if(candidateSaved && duplicateCandidate.length > 0){
+                logger.warn("Created duplicated candidate. Details:", candidateSaved);
+                return res.status(307).send({ warning: "Created duplicated candidate!" });
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: 307?

**Review by phamduylong — CHANGES_REQUESTED** [review]

File: routes/candidateRoutes.js [hunk 3]
```diff
@@ -121,13 +137,30 @@ candidateRouter.put('/update/id=:id', authorizedAdmin, async (req, res) => {
...
+        const imgUrl = RegExp(imgRegex).exec(findImgUrl.img);
+        const params = {
+            Bucket: BUCKET_NAME,
+            Key: imgUrl[0],
+        };
+        const command = new DeleteObjectCommand(params);
+        const imgDeleted = await s3.send(command);
+        if(!imgDeleted){
+            logger.error("Unable to delete image from s3");
+            return res.status(400).send({error: 'Unable to delete image'});
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: How is this a bad request? This is totally Amazon failing to delete the image, user did not make a bad request

File: routes/candidateRoutes.js [hunk 4]
```diff
@@ -85,33 +85,49 @@ candidateRouter.post('/create', authorizedAdmin, upload.single('file'), async (r
...
+            Key: imgUrl[0],
+            Body: image,
+            ContentType: 'image/jpeg',
+            ACL: 'public-read'
+        };
+        const command = new PutObjectCommand(params);
+        const imageSaved = await s3.send(command);
+        if(!imageSaved){
+            logger.error("Unable to upload image to s3");
+            return res.status(400).send({error: 'Unable to upload image'});
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: Again, bad request is not suitable here.

**Review by phamduylong — CHANGES_REQUESTED** [review]

File: routes/candidateRoutes.js (hunk 1 above)
Comment (kendrick-807) [reply routes/candidateRoutes.js]: Why?

File: routes/candidateRoutes.js (hunk 3 above)
Comment (kendrick-807) [reply routes/candidateRoutes.js]: Internal server error then?

File: routes/candidateRoutes.js (hunk 2 above)
Comment (kendrick-807) [reply routes/candidateRoutes.js]: Oh yeah lol

File: routes/candidateRoutes.js [hunk 5]
```diff
@@ -73,45 +73,60 @@ candidateRouter.post('/create', authorizedAdmin, upload.single('file'), async (r
...
+            Key: `${name}_${currentTimestamp}`,
+            Body: image.buffer,
+            ContentType: 'image/jpeg',
+            ACL: 'public-read'
+        };
+        const command = new PutObjectCommand(params);
+        const imageSaved = await s3.send(command);
+        if(!imageSaved){
+            logger.error("Unable to upload image to s3");
+            return res.status(400).send({error: 'Unable to upload image'});
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: This didn't change, did it?

File: routes/candidateRoutes.js [hunk 6]
```diff
@@ -121,13 +136,30 @@ candidateRouter.put('/update/id=:id', authorizedAdmin, async (req, res) => {
...
+        const imgUrl = RegExp(imgRegex).exec(findImgUrl.img);
+        const params = {
+            Bucket: BUCKET_NAME,
+            Key: imgUrl[0],
+        };
+        const command = new DeleteObjectCommand(params);
+        const imgDeleted = await s3.send(command);
+        if(!imgDeleted){
+            logger.error("Unable to delete image from s3");
+            return res.status(500).send({error: 'Unable to delete image'});
```
Comment (phamduylong) [inline routes/candidateRoutes.js:152]: Perhaps 424 is even better for all the Amazon failing cases. Check if it's an official status

File: routes/candidateRoutes.js [hunk 7]
```diff
@@ -121,13 +136,30 @@ candidateRouter.put('/update/id=:id', authorizedAdmin, async (req, res) => {
 
 candidateRouter.delete('/delete/id=:id', authorizedAdmin, async (req, res) => {
     const candidateId = req.params.id;
-    Candidate.deleteOne({ id: Number(candidateId) }).then(() => {
+    //Getting the image url from db before deleting it
+    const findImgUrl = await Candidate.findOne({ id: Number(candidateId) });
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: findImgUrl is a candidate?

**Review by phamduylong — CHANGES_REQUESTED** [review]

File: routes/candidateRoutes.js (hunk 6 above)
Comment (kendrick-807) [reply routes/candidateRoutes.js:152]: Dont think 424 is an official status code for amazon

Comment (kendrick-807) [reply routes/candidateRoutes.js:152]: I think 500 is fine for this case

**Comment by phamduylong** [comment]
Note for next time: not worth being lazy on this, it's not much typing each ticket
image

**Review by phamduylong — APPROVED** [review]

**Pull request merged**